eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -679,3 +631,145 @@ fromPrevGovActionIds (PrevGovActionIds pforest@(PForest _ _ _ _)) =
         & pfHardForkL . prRootL .~ pfHardFork
         & pfCommitteeL . prRootL .~ pfCommittee
         & pfConstitutionL . prRootL .~ pfConstitution
+
+---------------------
+-- Debugging tools --
+---------------------
+
+-- | Wraper type, which serves as a composition of @`Tree` . `StrictMaybe`@
+--
+-- Also its Show instance will print a nice tree structure.
+data TreeMaybe a = TreeMaybe {unTreeMaybe :: Tree (StrictMaybe a)}
+  deriving (Eq)
+
+instance Show (TreeMaybe (GovPurposeId p era)) where
+  show = ("\n" <>) . drawTree . fmap showGovPurposeId . unTreeMaybe
+    where
+      showGovPurposeId = \case
+        SNothing -> "x"
+        SJust (GovPurposeId govActionId) -> T.unpack (govActionIdToText govActionId)
+
+-- | Partial version of `toPForestEither`
+toPForest :: (Era era, HasCallStack) => Proposals era -> PForest TreeMaybe era
+toPForest = either error id . toPForestEither
+
+-- | Convert `Proposals` into a valid `Tree`
+toPForestEither :: Era era => Proposals era -> Either String (PForest TreeMaybe era)
+toPForestEither Proposals {pProps, pRoots, pGraph} = do
+  unless (OMap.invariantHolds' pProps) $ Left "OMap invariant is violated"
+
+  (pfPParamUpdate, nodesPParamUpdate) <-
+    toPTree (pfPParamUpdate pRoots) (unPGraph (pfPParamUpdate pGraph))
+  (pfHardFork, nodesHardFork) <-
+    toPTree (pfHardFork pRoots) (unPGraph (pfHardFork pGraph))
+  (pfCommittee, nodesCommittee) <-
+    toPTree (pfCommittee pRoots) (unPGraph (pfCommittee pGraph))
+  (pfConstitution, nodesConstitution) <-
+    toPTree (pfConstitution pRoots) (unPGraph (pfConstitution pGraph))
+
+  let allNodes =
+        Set.unions
+          [ Set.map unGovPurposeId nodesPParamUpdate
+          , Set.map unGovPurposeId nodesHardFork
+          , Set.map unGovPurposeId nodesCommittee
+          , Set.map unGovPurposeId nodesConstitution
+          ]
+      propsMap = OMap.toMap pProps
+      guardUnknown = do
+        let unknown = allNodes Set.\\ Map.keysSet propsMap
```

Comment:
`Set.difference` Return elements of the first set not existing in the second set, in that, it is not a commutative operation. Should we instead check that the union and intersection of the two sets is the same? We could even test directly for equality?

Reply:
Maybe ```(A `union` B) \\ (A `intersection` B)```

Reply:
No we should not. :smile:  

That would be the same wrong invariant that was in the previous implementation: 
https://github.com/IntersectMBO/cardano-ledger/blob/6fafb7269679cd32125e833ca70ca31f8d1bd4fc/eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs#L353-L354

Reply:
There is, however, one invariant that I just thought of that would be useful to implement that is related to this:

Namely, making sure that every govAction that exists in the `propsMap` that can have a parent, specifies the same parent in the `pRoots` V `pGraph`

I'll send you a commit once I implement it.

Reply:
> Maybe (A `union` B) \\ (A `intersection` B)

This won't work by itself.

```
(A `union` B) \\ (A `intersection` B) == (A \\ B) `union` (B \\ A)
```

In our case `(B \\ A)` is all of the governance actions that don't have parents (InfoAction and TreasuryWithdrawals), so we should not report them as unknown. It only is only `(A \\ B)` that are guaranteed to be invalid.

There is however part of `(B \\ A)` that can be invalid, we only need to check such B that has a parent. In which case, the check I suggest here will do the trick: https://github.com/IntersectMBO/cardano-ledger/pull/4016#discussion_r1469930329